### ISSUE TITLE
Lost notification when first block2 notification is not complete.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#211]
I faced a strange issue at client side when receiving 1 notification without blockmode during block2 transfer of the first notification on the same path. 

I will try to explain it but you'll see this is a bit tricky :p

1) Here is the failing scenario :
```
// client send a GET with observe on /test resource.
CON [MID=25457, T=f06bf010], GET, /test, (observe=0)    ----->

// server ACK the request and begin to send the response using block2
<-----   ACK [MID=25457], 0
<-----   CON [MID=111, T=f06bf010], 2.05, 2:0/1/1024, (observe=1)    

// client ACK and request the next block
ACK [MID=111], 0                          ----->
CON [MID=25458, T=ae04], GET, /test, 2:1/0/1024    ----->

// server finally send a new more recent notification but smaller without block2
<-----   CON [MID=222, T=f06bf010], 2.05, (observe=2)    

// we ack the response 
ACK [MID=222], 0                          ----->
// **** here at client side we get the response thanks to request.waitForResponse() : it works ****

// server send a new response and client acknowledge it
<-----   CON [MID=223, T=f06bf010], 2.05, (observe=3)  
ACK [MID=223], 0                          ----->
// **** But we never receive notification using Endpoint.addNotification() : failed :'( ****
```
2) I write a test about it [here](https://github.com/eclipse/californium/commit/fa4458c826c3e31a710014bb82c33d7213113242#diff-15e90779021001ad3493c536fd0165c0R315)

3) Here is a recap of how observe should works with notification:
We create a new exchange for new observe request then store it in Matcher maps. 
When we receive the response, we notify using `MessageObserver` API and we remove exchange from maps.
When a new response (notification) is received, [we create a new exchange from ObservationStore](https://github.com/eclipse/californium/blob/missing_notification/californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java#L191) to handle it and we notify using `NotificationListener` API.

4) What happens in our use case :
The maps is cleaned in [ExchangeObserverImpl.completed(Exchange)](https://github.com/eclipse/californium/blob/missing_notification/californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java#L294) using MID and Token of exchange.getCurrentRequest.
In our case when we receive the first none block2 notification, the currentRequest of the exchange is still the block2 request. So we don't clean the map correctly and we will never be able to recreate exchange from `ObservationStore` for new notification. That's why we never receive notification through `notificationListener`.

5) Solution :
I think when we receive a more recent notification, we should clean old block2 and matcher maps.
[Here](https://github.com/eclipse/californium/commit/8e83a1be9aeb8d4290e411113d0182ee23cbe224) is a quick fix which do that. Does it make sense to you ?

**Comment by sophokles73** [comment]
You seem to consider the response to the client's initial request as the first notification (which requires a block2 transfer). Then you argue that while this transfer is in progress, the server sends a notification which does not require a block2 transfer and you seem to suggest that the client should get the body of this notification in response to its initial request instead of the body being transferred using block2.
IMHO the first block2 transfer does not constitute a notification but simply is the response to the initial request. From the client's perspective, the observe relation is not even established until the full response (transferred using block2) is delivered to the observe layer. This is because the observe option that is set on the server's response can be interpreted by the observe layer only as part of the fully assembled response (because the observer layer is higher in the stack than the blockwise layer).
I therefore do not think that the response should contain the body of the notification but the body of the response being transferred via block2. This may mean that the first notification being sent by the server is NOT recognized as such by the client which, in this case, seems correct as well to me because, technically, the observe relation is not yet fully established ...

**Comment by sbernard31** [comment]
> Then you argue that while this transfer is in progress, the server sends a notification which does not require a block2 transfer and you seem to suggest that the client should get the body of this notification in response to its initial request instead of the body being transferred using block2.

Why I think that ?
1) if this new notification used block2 we would stop the previous one and handle the new one. And in this case we consider this new one as the first response. So why should we do differently with a none block2 scenario ?
2) It seems to be the CoAP observe spirit, only the more recent notification make sense.

**Comment by sophokles73** [comment]
> if this new notification used block2 we would stop the previous one and handle the new one. And in this case we consider this new one as the first response. So why should we do differently with a none block2 scenario?

If that is the current behavior then I think it is wrong for the same reason I have already laid out: the observe relation is not yet established.
ad 2) Again, if the observe relation were already established, I would agree. Since it is not, I don't :-)

**Comment by boaks** [comment]
> From the client's perspective, the observe relation is not even established until the full response (transferred using block2) is delivered to the observe layer. 

> the observe relation is not yet established.

Are these "artifacts" of the implementation?
Why must the blockwise transfer be finished? Especially, if the read data is deprecated.
Why sould the server wait for sending a notify, until the blockwise response was finished?

**Comment by sophokles73** [comment]
> Why must the blockwise transfer be finished? Especially, if the read data is deprecated.

My argumentation is based on the fact that the layer's above the `BlockwiseLayer` get the fully assembled response only after the block2 transfer is completed. The `ObserveLayer` is responsible for managing observe relations, thus the observe relation is only established once the block2 transfer is complete.

If we would opt for what Simon is proposing then we might end up in a situation where the client never receives a response to its initial request if the server sends notifications faster than the client can retrieve them (using a block2 download).

> Why should the server wait for sending a notify, until the blockwise response was finished?

The server doesn't need to wait. I think we are discussing whether the client needs to consider any notifications even before the observe relation has been established.

**Comment by boaks** [comment]
OK, so talking about "`...Layer`", I think, its the behaviour of the implementation.

> I think we are discussing whether the client needs to consider any notifications even before the observe relation has been established.

Makes sense for me.

**Comment by boaks** [comment]
> Considering notifications before the observe relation is established?

That part.

**Comment by sbernard31** [comment]
@sophokles73 How do you make the difference between the response which established the observe relation and a notification ?

Imagine this use case :
```
// client send a GET with observe on /test resource.
CON [MID=25457, T=f06bf010], GET, /test, (observe=0)    ----->
// server ACK the request.
<-----   ACK [MID=25457], 0

// then server send 2 responses at the same time
<-----   CON [MID=1212, T=f06bf010], 2.05, 2:0/1/1024, (observe=12)    
<-----   CON [MID=141, T=f06bf010], 2.05, 2:0/1/1024, (observe=33)    
// but because of UDP, client receive it like this 
<-----   CON [MID=141, T=f06bf010], 2.05, 2:0/1/1024, (observe=33)    
<-----   CON [MID=1212, T=f06bf010], 2.05, 2:0/1/1024, (observe=12)    
```

You will process the second notification as a response which established the observe relation.
How bad is that ?

**Comment by sophokles73** [comment]
> You will process the second notification[MID=141, observer=33] as a response which established the observe relation.

Yes, obviously, but we would also process it as the response if we implemented what you propose, because the notification [MID=1212, observe=12] would be discarded as outdated (12 < 33). There is nothing we can do about this problem, I guess, or can we?

**Comment by boaks** [comment]
> How bad is that ?

I german, I would read implicitly, that this **is bad**.
So, do you think it's bad, or do you ask, just to find out, if someone else thinks it's bad?

**Comment by boaks** [comment]
I tried to understood the current clientside behaviour of Exchange with Blockwise.
For me it seem, that the Exchange is just not completed for the callbacks, but I don't see that the Exchange is really used. Or did I oversee something?

Followup the first message, the issues seems to be, that notifications rely on missing Exchanges, and blockwise on a extended lifetime of that Exchange. What ends up in that conflict.

So, if we understand, why blockwise still needs the extended Exchange, we may get ideas how to fix it.

**Comment by sbernard31** [comment]
>So, do you think it's bad, or do you ask, just to find out, if someone else thinks it's bad?

*It was a question*, but the idea was just to prove there is not **one particular** notification which established the relation. I mean the observe relation is established when we receive **a** valid notification with observe option >0. 
 As we can received it in bad order, we just keep the more recent value. (we discard old notification)
In that case, I do not understand why it's bad to stop the bloc2 transfer and considering the observe relation as established.

We currently see the problem at server side, but *let's see it at server side :*
The server receive an observe request on a large resource (so it should use block2).
It starts to send the response using block2.
The value change and it want to notify the client, but this time the value is not so large (no need to block2).
It will just stop the block2 transfer as the value is not relevant anymore and send a new notification (without block2 as the value is smaller).

Does it still not make sense for you ?

**Comment by boaks** [comment]
> Does it still not make sense for you ?

Sure. (But the coap server doesn't stop the "block2 transfer", it may just not anser with content to a request for a block, which not longer existing.)

So I still think, the point is the usage of the Exchange, for the Observe (missing Exchange) and for the blockwise (sticky Exchange). So my feeling for the proper fix is, to remove the "Blockwise Exchange" from the "token map of the matcher". But I'm not sure, what this will cause then accidently.

**Comment by sophokles73** [comment]
I am getting the feeling that we are having an academic discussion. Is there a real problem or is this just about *optimizing* for an extremely rare case? As @sbernard31 pointed out, the observe relation is established, once the first response with an *observe* option > 0 reaches the `ObserveLayer`.  Because of the nature of UDP we have no control over which notification that will be (if the notifications are sent in sufficiently short succession). Once the observe relation is established, everything works as designed and expected, doesn't it?

**Comment by sbernard31** [comment]
This should be an extremely rare case but unfortunately we (at sierra) faced some devices which are expert in making hypothetic use cases real...

And the produce a real bug, because when this happens notification is not received anymore.

I mean :
block2 notification is never finished (why not)
first notification is handle as the first response (it's ok)
all other notification is lost in the CoAP Layer (I mean we never receive notification through `notificationListener`

**Comment by boaks** [comment]
As mentioned in #231, that PR should also fix this.
(At least it fixes the unit test provided in this issue.)

@sbernard31 
Could I contribute your test or do you do it on your own?
Do you have additional tests to see, if PR #231 fixes this issue?
Or is it fixed for you as well?

**Comment by sbernard31** [comment]
> Could I contribute your test or do you do it on your own?

You can contribute my test.

> Do you have additional tests to see, if PR #231 fixes this issue?

No.

> is it fixed for you as well?

It seems to be fixed. I still find my solution more understandable but as I failed to defend it,  I suppose it's ok like this.

**Comment by boaks** [comment]
Focusing on this issue only, your fix was clearly easier to understand! 
But being afraid of other similar "leaks" in the exchange management (e.g. for the special blockwise stuff),  I prefer that general cleanup of #231.

So, do you close the issue?

**Comment by sbernard31** [comment]
Yep.
Fixed by #231.
